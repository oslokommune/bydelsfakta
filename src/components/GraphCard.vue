<template>
  <section class="card-container" :class="{ large: settings.size === 'large' }">
    <div class="card" :class="{ fullscreen }" @keydown.escape="toggleFullscreen" :tabindex="fullscreen ? 0 : false">
      <button
        v-if="fullscreen"
        role="menuitem"
        @click="toggleFullscreen"
        @keyup.enter="toggleFullscreen"
        class="close-fullscreen"
        tabindex="0"
        :title="fullscreen ? $t('graphCard.fullscreen.exit.aria') : $t('graphCard.fullscreen.open.aria')"
        :aria-label="fullscreen ? $t('graphCard.fullscreen.exit.aria') : $t('graphCard.fullscreen.open.aria')"
      >
        <ok-icon icon-ref="fullscreenExit" :options="{ size: 'small' }"></ok-icon>
      </button>
      <header class="card__header">
        <div class="card__headertext">
          <h2 class="card__title">{{ settings.heading }}</h2>
        </div>
        <nav class="card__nav">
          <div class="tabs" ref="tabsRef" role="tablist">
            <resize-observer @notify="showTabsOrSelect"></resize-observer>
            <template v-for="(tab, index) in settings.tabs">
              <button
                :disabled="mode === 'map' || mode === 'about'"
                ref="tabRef"
                role="tab"
                :aria-label="tab.label"
                :id="`tabButton-${settings.heading}-${index}`"
                :key="index"
                @click="activeTab(index)"
                :class="{ active: active === index, 'tabs__tab--hidden': !showAsTabs }"
                class="tabs__tab"
                v-text="tab.label"
                v-if="
                  tab.production === productionMode && tab.production !== null
                    ? true
                    : tab.production === true && productionMode === false
                    ? true
                    : productionMode === null
                "
              ></button>
            </template>
            <select v-if="!showAsTabs" v-model="active" class="select tabs__select" aria-hidden>
              <option
                :value="index"
                v-for="(tab, index) in settings.tabs"
                v-text="tab.label"
                :key="'tab' + tab.label"
              ></option>
            </select>
          </div>

          <div class="card__toggle-menu">
            <button
              class="card__toggle-button"
              @click="mode = 'graph'"
              :class="{ 'card__toggle-button--active': mode === 'graph' }"
              :aria-label="$t('graphCard.mode.graph')"
              :title="$t('graphCard.mode.graph')"
            >
              <ok-icon
                class="context-menu__dropdown-item-icon"
                icon-ref="graph"
                stroke-width="2.5"
                :options="{ size: 'tiny' }"
              ></ok-icon>
            </button>
            <button
              class="card__toggle-button"
              @click="mode = 'table'"
              :class="{ 'card__toggle-button--active': mode === 'table' }"
              :aria-label="$t('graphCard.mode.table')"
              :title="$t('graphCard.mode.table')"
            >
              <ok-icon
                class="context-menu__dropdown-item-icon"
                icon-ref="data-table"
                stroke-width="2.5"
                :options="{ size: 'tiny' }"
              ></ok-icon>
            </button>

            <button
              v-if="settings.map"
              class="card__toggle-button"
              @click="mode = 'map'"
              :class="{ 'card__toggle-button--active': mode === 'map' }"
              :aria-label="$t('graphCard.mode.map')"
              :title="$t('graphCard.mode.map')"
            >
              <ok-icon
                class="context-menu__dropdown-item-icon"
                icon-ref="map-pin"
                stroke-width="2.5"
                :options="{ size: 'tiny' }"
              ></ok-icon>
            </button>
          </div>
          <div @keydown.escape="closeMenu()" v-click-outside="closeMenu" class="context-menu">
            <button
              class="context-menu__button"
              :class="{ 'card__toggle-button--active': showDropdown }"
              @click="showDropdown = !showDropdown"
              aria-haspopup="true"
              :aria-label="showDropdown ? $t('graphCard.dropdown.close') : $t('graphCard.dropdown.open')"
              id="context-menu-button"
              :title="showDropdown ? $t('graphCard.dropdown.close') : $t('graphCard.dropdown.open')"
            >
              <ok-icon icon-ref="hamburger"></ok-icon>
              <span class="button-label">Valg</span>
            </button>
            <div v-if="showDropdown" class="context-menu__dropdown" role="menu">
              <button
                role="menuitem"
                @click="toggleFullscreen"
                @keyup.enter="toggleFullscreen"
                class="context-menu__dropdown-item"
                tabindex="0"
                :title="fullscreen ? $t('graphCard.fullscreen.exit.aria') : $t('graphCard.fullscreen.open.aria')"
                :aria-label="fullscreen ? $t('graphCard.fullscreen.exit.aria') : $t('graphCard.fullscreen.open.aria')"
              >
                <ok-icon
                  :icon-ref="fullscreen ? 'fullscreenExit' : 'fullscreen'"
                  :options="{ size: 'small' }"
                ></ok-icon>
                <span>{{
                  fullscreen ? $t('graphCard.fullscreen.exit.label') : $t('graphCard.fullscreen.open.label')
                }}</span>
              </button>

              <button
                role="menuitem"
                @click="
                  mode === 'about' ? (mode = 'graph') : (mode = 'about');
                  showDropdown = false;
                "
                @keyup.enter="
                  mode === 'about' ? (mode = 'graph') : (mode = 'about');
                  showDropdown = false;
                "
                class="context-menu__dropdown-item"
                tabindex="0"
                :title="$t('graphCard.about.aria')"
                :aria-label="$t('graphCard.about.aria')"
                id="context-menu-button-about"
              >
                <ok-icon icon-ref="help" :options="{ size: 'small' }"></ok-icon>
                <span>{{ $t('graphCard.about.label') }}</span>
              </button>

              <button
                role="menuitem"
                :disabled="mode !== 'graph'"
                @click="savePng(settings.tabs[active].id)"
                @keyup.enter="saveSvg(settings.tabs[active].id)"
                class="context-menu__dropdown-item"
                tabindex="0"
                :title="$t('graphCard.savePNG.aria')"
                :aria-label="$t('graphCard.savePNG.aria')"
                id="context-menu-button-png"
              >
                <ok-icon icon-ref="photo" :options="{ size: 'small' }"></ok-icon>
                <span>{{ $t('graphCard.savePNG.label') }}</span>
              </button>
              <button
                role="menuitem"
                :disabled="mode !== 'graph' || ie11"
                :title="ie11 ? $t('ie11.disabled') : $t('graphCard.saveSVG.aria')"
                class="context-menu__dropdown-item"
                :aria-label="$t('graphCard.saveSVG.aria')"
                tabindex="0"
                @click="saveSvg(settings.tabs[active].id)"
                @keyup.enter="saveSvg(settings.tabs[active].id)"
                id="context-menu-button-svg"
              >
                <ok-icon icon-ref="photo" :options="{ size: 'small' }"></ok-icon>
                <span>{{ $t('graphCard.saveSVG.label') }}</span>
              </button>

              <button
                role="menuitem"
                :disabled="mode === 'map'"
                class="context-menu__dropdown-item"
                :title="$t('graphCard.saveCSV.aria')"
                :aria-label="$t('graphCard.saveCSV.aria')"
                tabindex="0"
                @click="saveCsv()"
                @keyup.enter="saveCsv()"
                id="context-menu-button-csv"
              >
                <ok-icon icon-ref="download" :options="{ size: 'small' }"></ok-icon>
                <span>{{ $t('graphCard.saveCSV.label') }}</span>
              </button>
            </div>
          </div>
        </nav>
      </header>
      <graph-instance
        @updateDate="setDate"
        v-if="settings.tabs[active] !== undefined && (mode === 'graph' || mode === 'table')"
        :settings="settings.tabs[active]"
        :sources="settings.sources"
        :mode="mode"
        ref="graph"
      />
      <div class="map-container" v-if="mode === 'map'">
        <v-leaflet
          v-if="settings.map"
          :district="geoDistricts"
          :settings="settings.map"
          :data-url="`${settings.map.url}?geography=${district}`"
        ></v-leaflet>
      </div>
      <div class="about-container" v-if="mode === 'about'">
        <h3>{{ $t('graphCard.about.label') }}</h3>

        <p v-if="date">{{ $t('graphCard.about.updated') }}: {{ date }}</p>
        <p v-if="settings.about" v-html="settings.about"></p>
        <p v-if="settings.sources">
          {{ settings.sources.length > 1 ? $t('graphCard.sources.plural') : $t('graphCard.sources.singular') }}:
          <span v-for="(source, i) in settings.sources" :key="'source_' + i">
            <a :href="source.url">{{ source.name }}</a>
            <span v-if="i < settings.sources.length - 1">, </span>
          </span>
        </p>
      </div>
    </div>
  </section>
</template>

<script>
import { mapGetters, mapState } from 'vuex';
import * as d3 from 'd3';
import GraphInstance from './GraphInstance';
import downloadSvg from '../util/downloadSvg';
import downloadPng from '../util/downloadPng';
import tableToCsv from '../util/tableToCsv';
import VLeaflet from './VLeaflet';
import OkIcon from './OkIcon';

export default {
  name: 'GraphCard',
  components: { GraphInstance, VLeaflet, OkIcon },
  data() {
    return {
      active: 0,
      showDropdown: false,
      mode: 'graph',
      date: '',
      fullscreen: false,
      showAsTabs: true,
    };
  },
  computed: {
    ...mapState(['districts', 'ie11', 'productionMode']),
    ...mapGetters(['geoDistricts']),
    district() {
      if (this.districts[0] === 'alle') return '00';
      if (this.districts.length === 1) return this.districts[0];
      return '00';
    },
  },
  props: {
    settings: {
      type: Object,
      required: true,
    },
  },

  methods: {
    setDate(dateStr) {
      if (!dateStr) {
        this.date = 'Ikke tilgjengelig';
        return;
      }
      const parseTime = d3.timeParse('%Y-%m-%d');
      const formatTime = d3.timeFormat('%x');
      this.date = formatTime(parseTime(dateStr));
    },

    toggleFullscreen() {
      const body = document.querySelector('body');
      this.showDropdown = false;

      if (this.fullscreen) {
        this.fullscreen = false;
        body.style.height = 'auto';
        body.style.overflow = 'auto';
      } else {
        this.fullscreen = true;
        body.style.height = '100vh';
        body.style.overflow = 'hidden';
      }
    },

    closeMenu() {
      if (this.showDropdown) {
        this.showDropdown = false;
      }
    },

    toggleShowTable() {
      this.showTable = !this.showTable;
      this.showDropdown = false;
    },

    activeTab(index) {
      this.active = index;
    },

    saveSvg(id) {
      const filename = `${this.$route.params.district}_${id}`;
      const svgData = this.$refs.graph.$refs.svg.outerHTML;
      downloadSvg(svgData, filename);
      this.closeMenu();
    },

    savePng(id) {
      const filename = `${this.$route.params.district}_${id}.png`;
      downloadPng(this.$refs.graph.$refs.svg, filename);

      this.closeMenu();
    },

    saveCsv() {
      tableToCsv(this.$refs.graph.$refs.tableContainer.querySelector('table'));
      this.closeMenu();
    },

    // Sets 'showAsTabs' true/false by comparing the accumulated width
    // of all the tabs with the available space
    showTabsOrSelect() {
      let tabsWidth = 0;
      const tabs = this.$refs.tabRef;
      tabs.forEach(tab => {
        tabsWidth += tab.offsetWidth;
      });
      this.showAsTabs = this.$refs.tabsRef.offsetWidth > tabsWidth;
    },
  },

  watch: {
    '$route.params.topic'() {
      this.active = 0;
      this.mode = 'graph';
      this.showTabsOrSelect();
    },
  },
};
</script>

<style scoped lang="scss">
@import './../styles/colors';
@import './../styles/variables';

.map-container,
.about-container {
  height: 560px;
}

.about-container {
  padding: 1rem;
}

.card-container {
  flex: 50% 1 10;
  padding: 1em;
  width: 100%;

  @media screen and (min-width: $break-md) {
    max-width: 50%;
    width: 50%;
  }

  &.large {
    flex-basis: 100%;
    max-width: 100%;
  }
}

.close-fullscreen {
  align-items: center;
  background: rgba(black, 0.1);
  border-radius: 50%;
  display: flex;
  justify-content: center;
  padding: 0.5rem;
  position: absolute;
  right: 0.5rem;
  top: 0.5rem;

  &:hover {
    background: rgba(black, 0.2);
  }
}

.card {
  background: white;
  border: 1px solid $color-border;
  min-height: 18em;
  width: 100%;

  &.fullscreen {
    bottom: 1rem;
    height: calc(100vh - 2rem);
    left: 1rem;
    outline: 2rem solid rgba(black, 0.9);
    position: fixed;
    right: 1rem;
    top: 1rem;
    width: calc(100vw - 2rem);
    z-index: 5;
  }

  &__header {
    border-bottom: 1px solid $color-border;
    display: flex;
    flex-direction: column;
    width: 100%;
    z-index: 1;
  }

  &__headertext {
    align-items: baseline;
    display: flex;
    padding: 0 1rem;
  }

  &__title {
    color: $color-purple;
    font-size: $font-large;
    font-weight: 500;
    margin-bottom: 0.5rem;
    margin-right: 1rem;
  }

  &__published {
    color: rgba($color-purple, 0.65);
    font-size: $font-small;
    text-transform: uppercase;
  }

  &__nav {
    align-items: center;
    display: flex;
    padding-left: 1rem;
    padding-right: 0;
    position: relative;
  }

  &__toggle-menu {
    display: flex;
    margin-left: auto;
    padding-left: 1rem;
  }

  &__toggle-button {
    align-items: center;
    border-radius: 50%;
    cursor: pointer;
    display: flex;
    height: 2.5em;
    justify-content: center;
    position: relative;
    width: 2.5em;

    &::before {
      background: $color-blue;
      border-radius: 50%;
      bottom: 0;
      content: '';
      left: 0;
      position: absolute;
      right: 0;
      top: 0;
      transform: scale(1);
      transition: all 0.3s cubic-bezier(0.25, 0, 0, 1);
      z-index: 0;
    }

    &::after {
      background: white;
      border-radius: 50%;
      bottom: 0;
      content: '';
      left: 0;
      position: absolute;
      right: 0;
      top: 0;
      transform: scale(1);
      transition: all 0.3s cubic-bezier(0.25, 0, 0, 1);
      z-index: 0;
    }

    & > i {
      user-select: none;
      z-index: 1;
    }

    &:hover:not(&--active)::after {
      // background-color: rgba($color-border, 0.35);
      transform: scale(0.8);
      transition: all 0.1s cubic-bezier(0.3, 0, 0.5, 1);
    }

    &--active::before {
      transform: scale(1);
      transition: all 0.5s cubic-bezier(0.3, 0, 0.5, 1);
    }

    &--active::after {
      transform: scale(0);
      transition: all 0.5s cubic-bezier(0.3, 0, 0.5, 1);
    }
  }
}

/* Style the tabs */
.tabs {
  display: flex;
  flex-grow: 1;
  height: 4rem;
  overflow-x: hidden;
  padding: 4px 0;
  position: relative;

  &__select {
    height: 3rem;
    left: 0;
    margin-right: 1rem;
    position: absolute;
    right: 1rem;
    top: 0.5rem;
  }

  &__tab {
    color: $color-purple;
    cursor: pointer;
    font-weight: 500;
    padding: 0.75em;
    transition: background-color 0.2s;
    white-space: nowrap;

    &--hidden {
      height: 0;
      opacity: 0;
    }

    &[disabled]:not(.tabs__tab--hidden),
    &.active[disabled]:not(.tabs__tab--hidden) {
      color: rgba($color-purple, 0.35);
      cursor: default;
      opacity: 0.8;

      &::after {
        opacity: 0;
      }
    }

    &.active {
      cursor: default;
      position: relative;

      &::after {
        background-color: $color-blue;
        bottom: -4px;
        content: '';
        height: 5px;
        left: 0;
        position: absolute;
        right: 0;
      }
    }

    /* Change background color of tabs on hover */
    &:hover:not(.active):not(:disabled) {
      background-color: rgba($color-grey-100, 0.5);
      position: relative;
    }
  }
}

.context-menu {
  margin-left: auto;
  position: relative;
  z-index: 4;

  &__button {
    align-items: center;
    border-radius: 0 !important;
    cursor: pointer;
    display: flex;
    height: 3rem;
    padding-left: 0.65rem;
    padding-right: 0.65rem;

    &.card__toggle-button--active {
      background-color: $color-purple;
      color: $color-blue;
    }

    @media screen and (min-width: $break-md) {
      background-color: rgba($color-border, 0);
      height: 3rem;
      width: auto;

      &:hover:not(.card__toggle-button--active) {
        background-color: rgba($color-border, 0.35);
      }
    }

    .button-label {
      display: inline-block;
      font-weight: 500;
      margin-left: 0.25rem;
      text-transform: uppercase;
    }
  }

  &__dropdown {
    background-color: $color-purple;
    box-shadow: 0 2px 2px 0 $color-grey-200;
    position: absolute;
    right: 0;
    width: 310px;
    z-index: 1;

    &-item {
      align-items: center;
      color: $color-blue;
      cursor: pointer;
      display: flex;
      flex-direction: row;
      font-weight: 500;
      padding: 1.25rem 0.75rem;
      width: 100%;

      &:disabled {
        text-decoration: line-through;
      }

      &:hover:not(:disabled) {
        background-color: darken($color-purple, 10%);
      }

      &-icon {
        color: $color-purple;
        z-index: 1;
      }

      span {
        margin-left: 0.5rem;
        text-align: left;
      }
    }
  }
}
</style>
