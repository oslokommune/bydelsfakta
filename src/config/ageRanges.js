export default [
  { range: '', label: '-- Velg --', selected: true, disabled: true },
  { range: '[0, 1]', label: '0–1 år' },
  { range: '[2, 5]', label: '2–5 år' },
  { range: '[6, 12]', label: '6–12 år' },
  { range: '[13, 15]', label: '13–15 år' },
  { range: '[16, 18]', label: '16–18 år' },
  { range: '[19, 29]', label: '19–29 år' },
  { range: '[30, 39]', label: '30–39 år' },
  { range: '[40, 49]', label: '40–49 år' },
  { range: '[50, 59]', label: '50–59 år' },
  { range: '[60, 66]', label: '60–66 år' },
  { range: '[67, 79]', label: '67–79 år' },
  { range: '[80, 89]', label: '80–89 år' },
  { range: '[90, 119]', label: '90+ år' },
];
