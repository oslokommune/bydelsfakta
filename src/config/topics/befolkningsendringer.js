import { baseUrl } from '../../util/config';

export default {
  text: 'Fødte, døde og flytting',
  value: 'befolkningsendringer',
  production: null,
  cards: [
    {
      size: 'large',
      heading: 'Flytting etter alder',
      production: null,
      tabs: [
        {
          label: 'Status',
          method: 'value',
          heading: 'Flytting etter alder og innvandringskategori',
          template: 'm',
          url: `${baseUrl}/mockDataFlytting.json`,
          production: null,
        },
      ],
    },
  ],
  options: {
    kategori: 'Befolkning',
    tema: 'Fødte, døde og flytting',
    bgImage: `${baseUrl}/img/flytting`,
    txtColor: 'rgb(245, 173, 165)',
  },
  related: ['alder', 'befolkningsutvikling', 'innvandrerbefolkningen'],
};
