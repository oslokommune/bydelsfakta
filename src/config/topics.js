import alder from './topics/alder';
import befolkningsendringer from './topics/befolkningsendringer';
import boligpriser from './topics/boligpriser';
import bygningstyper from './topics/bygningstyper';
import eierform from './topics/eierform';
import befolkningsutvikling from './topics/befolkningsutvikling';
import husholdninger from './topics/husholdninger';
import innvandrerbefolkningen from './topics/innvandrerbefolkningen';
import levekaar from './topics/levekaar';
import trangboddhet from './topics/trangboddhet';

export const topicNames = [
  'befolkningsutvikling',
  'befolkningsendringer',
  'alder',
  'innvandrerbefolkningen',
  'husholdninger',
  'levekaar',
  'eierform',
  'bygningstyper',
  'boligpriser',
  'trangboddhet',
];

export const disabledTopics = ['befolkningsendringer'];

export const categories = [
  {
    kategori: 'Befolkning',
    color: 'rgb(182, 63, 50)',
    links: ['befolkningsutvikling', 'befolkningsendringer', 'alder', 'innvandrerbefolkningen', 'husholdninger'],
  },
  {
    kategori: 'Boforhold',
    color: 'rgb(27, 173, 120)',
    links: ['trangboddhet', 'eierform', 'bygningstyper', 'boligpriser'],
  },
  {
    kategori: 'Levek√•r',
    color: 'rgb(219, 160, 52)',
    links: ['levekaar'],
  },
];

export const topics = {
  alder,
  befolkningsendringer,
  boligpriser,
  bygningstyper,
  eierform,
  befolkningsutvikling,
  husholdninger,
  innvandrerbefolkningen,
  levekaar,
  trangboddhet,
};
