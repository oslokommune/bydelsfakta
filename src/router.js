import Vue from 'vue';
import Router from 'vue-router';

import { topicNames, disabledTopics } from './config/topics';
import allDistricts from './config/allDistricts';

const District = () => import('./views/District');
const Topic = () => import('./views/Topic');
const NotFound = () => import('./views/NotFound');

Vue.use(Router);

export const routes = [
  {
    path: '/',
    name: 'Home',
    component: District,
  },
  {
    path: '/bydel/:district?',
    name: 'District',
    component: District,
    props: true,
  },
  {
    path: '/bydel/:district/:topic',
    name: 'Topic',
    component: Topic,
    props: true,
  },
  {
    path: '*',
    name: 'NotFound',
    component: NotFound,
  },
];

const router = new Router({
  mode: 'history',
  routes,
});

// Temp while we find out what our homepage should look like
router.beforeEach((to, from, next) => {
  if (to.params.topic) {
    if (!topicNames.find(name => name === to.params.topic) || disabledTopics.includes(to.params.topic)) {
      next({ name: 'NotFound', params: [to.path] });
    }
  }

  if (to.params.district) {
    const districts = to.params.district.split('-');
    if (districts[0] === 'alle') {
      next();
    } else if (districts.length === 1) {
      if (allDistricts.find(district => district.uri === districts[0])) {
        next();
      } else if (allDistricts.find(district => district.key === districts[0])) {
        next();
      } else {
        next({ name: 'NotFound', params: [to.path] });
      }
    } else if (districts.length > 1) {
      const errors = [];
      districts.forEach(district => {
        if (!allDistricts.find(item => item.key === district)) {
          errors.push(district);
        }
      });
      if (errors.length > 0) {
        next({ name: 'NotFound', params: [to.path] });
      }
    }
  }

  if (to.params.topic !== from.params.topic || (to.name === 'Topic' && from.name === 'District')) {
    window.scroll(0, 0);
  }

  next();
});

export default router;
