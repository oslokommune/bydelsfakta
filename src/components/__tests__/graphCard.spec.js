import 'whatwg-fetch';
import { mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import VueRouter from 'vue-router';
import vueResize from 'vue-resize';
import VueAnalytics from 'vue-analytics';
import GraphCard from '../GraphCard';
import clickOutside from '../../directives/clickOutside';
import { topics } from '../../config/topics';
import mockStore from '../../../tests/MockStore';
import setupI18n from '../../i18n';
import { routes } from '../../router';

const i18n = setupI18n();

describe('GraphCard', () => {
  let wrapper = null;
  let router = null;
  let store = null;

  beforeEach(() => {
    const localVue = createLocalVue();
    localVue.use(vueResize);
    localVue.use(Vuex);
    localVue.use(VueAnalytics, {
      id: 'UA-1234-5',
    });
    localVue.directive('click-outside', clickOutside);
    store = new Vuex.Store(mockStore);
    router = new VueRouter({ routes });
    wrapper = mount(GraphCard, {
      propsData: {
        settings: topics['alder'].cards[0],
      },
      localVue,
      router,
      i18n,
      store,
      stubs: {
        'ok-icon': true,
        spinner: true,
      },
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('renders graphCard-component and finds card-container-class', () => {
    expect(wrapper.classes('card-container')).toBe(true);
  });

  test('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test('change an active tab', () => {
    wrapper.setProps({ settings: topics['eierform'].cards[0] });
    wrapper.vm.activeTab(1);
    expect(wrapper.vm.active).toEqual(1);
    expect(wrapper.element).toMatchSnapshot();
  });

  test('change active tab multiple times', () => {
    wrapper.setProps({ settings: topics['husholdninger'].cards[0] });
    expect(wrapper.vm.active).toEqual(0);
    wrapper.vm.activeTab(1);
    expect(wrapper.vm.active).toEqual(1);
    wrapper.vm.activeTab(2);
    expect(wrapper.vm.active).toEqual(2);
  });

  test('if dropdown is false, stay false', () => {
    expect(wrapper.vm.showDropdown).toEqual(false);
    wrapper.vm.closeMenu();
    expect(wrapper.vm.showDropdown).toEqual(false);
  });

  test('if dropdown is true, change to false', () => {
    wrapper.setData({ showDropdown: true });
    expect(wrapper.vm.showDropdown).toEqual(true);
    wrapper.vm.closeMenu();
    expect(wrapper.vm.showDropdown).toEqual(false);
  });

  test('saveSvg is called once', () => {
    const spy = jest.spyOn(wrapper.vm, 'saveSvg');
    spy(wrapper.vm.settings.tabs[wrapper.vm.active].id);
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
