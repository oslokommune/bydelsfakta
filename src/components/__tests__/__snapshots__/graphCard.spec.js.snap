// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphCard change an active tab 1`] = `
<section
  class="card-container large"
>
  <div
    class="card"
  >
    <!---->
     
    <header
      class="card__header"
    >
      <div
        class="card__headertext"
      >
        <h2
          class="card__title"
        >
          Eierform
        </h2>
      </div>
       
      <nav
        class="card__nav"
      >
        <div
          class="tabs"
          role="tablist"
        >
          <div
            class="resize-observer"
            data-v-b329ee4c=""
            tabindex="-1"
          >
            <object
              aria-hidden="true"
              data="about:blank"
              tabindex="-1"
              type="text/html"
            />
          </div>
           
          <button
            aria-label="Andel"
            class="tabs__tab"
            data-id="tabButton-Eierform-0"
            role="tab"
          >
            Andel
          </button>
          <button
            aria-label="Matrise"
            class="tabs__tab active"
            data-id="tabButton-Eierform-1"
            role="tab"
          >
            Matrise
          </button>
          <button
            aria-label="Historisk"
            class="tabs__tab"
            data-id="tabButton-Eierform-2"
            role="tab"
          >
            Historisk
          </button>
           
          <!---->
        </div>
         
        <div
          class="card__toggle-menu"
        >
          <button
            aria-label="Graf"
            class="card__toggle-button card__toggle-button--active"
            title="Graf"
          >
            <ok-icon-stub
              class="context-menu__dropdown-item-icon"
              iconref="graph"
              options="[object Object]"
              stroke-width="2.5"
            />
          </button>
           
          <button
            aria-label="Tabell"
            class="card__toggle-button"
            title="Tabell"
          >
            <ok-icon-stub
              class="context-menu__dropdown-item-icon"
              iconref="data-table"
              options="[object Object]"
              stroke-width="2.5"
            />
          </button>
           
          <button
            aria-label="Kart"
            class="card__toggle-button"
            title="Kart"
          >
            <ok-icon-stub
              class="context-menu__dropdown-item-icon"
              iconref="map-pin"
              options="[object Object]"
              stroke-width="2.5"
            />
          </button>
        </div>
         
        <div
          class="context-menu"
        >
          <button
            aria-haspopup="true"
            aria-label="Åpne meny"
            class="context-menu__button"
            data-id="context-menu-button"
            title="Åpne meny"
          >
            <ok-icon-stub
              iconref="hamburger"
              options="[object Object]"
            />
             
            <span
              class="button-label"
            >
              Valg
            </span>
          </button>
           
          <!---->
        </div>
      </nav>
    </header>
     
    <div
      class="graph__shadow"
      role="tabpanel"
    >
      <div
        aria-hidden="true"
        class="graph__container"
      >
        <div
          class="spinner"
        >
          <spinner-stub />
           
          <span
            class="spinner__text"
          >
            Henter data
          </span>
        </div>
         
        <!---->
         
        <svg
          aria-hidden="true"
          class="graph__svg loading"
        />
      </div>
       
      <div
        class="graph__tablecontainer visually-hidden"
      >
        <h3
          class="table-heading"
        />
         
        <table
          class=""
        >
          <thead />
           
          <tbody />
        </table>
         
        <!---->
      </div>
       
      <div
        class="resize-observer"
        data-v-b329ee4c=""
        tabindex="-1"
      >
        <object
          aria-hidden="true"
          data="about:blank"
          tabindex="-1"
          type="text/html"
        />
      </div>
       
      <!---->
    </div>
     
    <!---->
     
    <!---->
  </div>
</section>
`;

exports[`GraphCard renders correctly 1`] = `
<section
  class="card-container large"
>
  <div
    class="card"
  >
    <!---->
     
    <header
      class="card__header"
    >
      <div
        class="card__headertext"
      >
        <h2
          class="card__title"
        >
          Befolkningen etter alder
        </h2>
      </div>
       
      <nav
        class="card__nav"
      >
        <div
          class="tabs"
          role="tablist"
        >
          <div
            class="resize-observer"
            data-v-b329ee4c=""
            tabindex="-1"
          >
            <object
              aria-hidden="true"
              data="about:blank"
              tabindex="-1"
              type="text/html"
            />
          </div>
           
          <button
            aria-label="Antall"
            class="tabs__tab active"
            data-id="tabButton-Befolkningen etter alder-0"
            role="tab"
          >
            Antall
          </button>
          <button
            aria-label="Andel"
            class="tabs__tab"
            data-id="tabButton-Befolkningen etter alder-1"
            role="tab"
          >
            Andel
          </button>
           
          <!---->
        </div>
         
        <div
          class="card__toggle-menu"
        >
          <button
            aria-label="Graf"
            class="card__toggle-button card__toggle-button--active"
            title="Graf"
          >
            <ok-icon-stub
              class="context-menu__dropdown-item-icon"
              iconref="graph"
              options="[object Object]"
              stroke-width="2.5"
            />
          </button>
           
          <button
            aria-label="Tabell"
            class="card__toggle-button"
            title="Tabell"
          >
            <ok-icon-stub
              class="context-menu__dropdown-item-icon"
              iconref="data-table"
              options="[object Object]"
              stroke-width="2.5"
            />
          </button>
           
          <!---->
        </div>
         
        <div
          class="context-menu"
        >
          <button
            aria-haspopup="true"
            aria-label="Åpne meny"
            class="context-menu__button"
            data-id="context-menu-button"
            title="Åpne meny"
          >
            <ok-icon-stub
              iconref="hamburger"
              options="[object Object]"
            />
             
            <span
              class="button-label"
            >
              Valg
            </span>
          </button>
           
          <!---->
        </div>
      </nav>
    </header>
     
    <div
      class="graph__shadow"
      role="tabpanel"
    >
      <div
        aria-hidden="true"
        class="graph__container"
      >
        <div
          class="spinner"
        >
          <spinner-stub />
           
          <span
            class="spinner__text"
          >
            Henter data
          </span>
        </div>
         
        <!---->
         
        <svg
          aria-hidden="true"
          class="graph__svg loading"
        />
      </div>
       
      <div
        class="graph__tablecontainer visually-hidden"
      >
        <h3
          class="table-heading"
        />
         
        <table
          class=""
        >
          <thead />
           
          <tbody />
        </table>
         
        <!---->
      </div>
       
      <div
        class="resize-observer"
        data-v-b329ee4c=""
        tabindex="-1"
      >
        <object
          aria-hidden="true"
          data="about:blank"
          tabindex="-1"
          type="text/html"
        />
      </div>
       
      <div
        class="help"
      >
        <button
          aria-label="Vis hjelpetekst"
          class="help__button"
        >
          <ok-icon-stub
            iconref="help"
            options="[object Object]"
          />
        </button>
         
        <!---->
      </div>
    </div>
     
    <!---->
     
    <!---->
  </div>
</section>
`;
